src/zql/ivm/graph/operators/distinct-operator.ts

```diff
@@ -0,0 +1,76 @@
+import {Entity} from '../../../../generate.js';
+import {Entry, Multiset} from '../../multiset.js';
+import {Version} from '../../types.js';
+import {DifferenceStream} from '../difference-stream.js';
+import {UnaryOperator} from './unary-operator.js';
+
+/**
+ * A dataflow operator that has a single input edge and a single output edge.
+ * It collapses multiple updates in the same transaction (version) to a single output.
+ * The multiplicity of the output entries is -1, 0, or 1 after the tx is done.
+ *
+ * The id property iu used to identify the entity.
+ */
+export class DistinctOperator<T extends Entity> extends UnaryOperator<T, T> {
+  // The entries for this version. The string is the ID of the entity.
+  readonly #entriesCache = new Map<Version, Map<string, Entry<T>>>();
+  // The emitted data for this version. The string is the ID of the entity. The
+  // number is the multiplicity outputted so far in this tx.
+  readonly #emitted = new Map<Version, Map<string, number>>();
+
+  constructor(input: DifferenceStream<T>, output: DifferenceStream<T>) {
+    super(input, output, (version, data) => this.#handleDiff(version, data));
+  }
+
+  #handleDiff(version: number, multiset: Multiset<T>): Multiset<T> {
+    // Clear data for old versions.
+    clearOldVersions(this.#entriesCache, version);
```

Comment:
I thought that each 'version' would call `handleDiff` one time only. If so, couldn't the cache just be a local?

Reply:
Maybe. I'm trying to understand if that is the case.

@tantaman ?

Reply:
> I thought that each 'version' would call handleDiff one time only.

Not now that we optimistically send data down as soon as we get it (no waiting at operators for inputs on both edges).

![image](https://github.com/rocicorp/rails/assets/1009003/4a3d51bd-4aee-470a-91d2-b2fd9e196fa5)

add will notify depth first down the left then the right.